### PR TITLE
Add API Methods to EnergyNet

### PR DESCRIPTION
**Pull request by JustAHuman-xD**
## Description
I am making this pull request to add 3 API methods to the EnergyNet class.
This would allow either Core Slimefun or Addons to access what is a part of a given energy network allowing for the addition of multiple approved suggestions as well as QOL additions.

## Proposed changes
Added three getter methods that return an unmodifiable version of the three custom Maps available in the EnergyNet class. Those being, getGenerators, getCapacitors, getConsumers.
Fixed a typo in the java docs.
Annotated getId to be Nonnull as it was not already

## Related Issues (if applicable)
There are no Related issues however these API changes can be used to implement, for example, https://discord.com/channels/565557184348422174/627048923122499584/1068050940294340608

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 1]
```diff
@@ -1,20 +1,5 @@
 package io.github.thebusybiscuit.slimefun4.core.networks.energy;
 
-import java.util.HashMap;
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:3]: You’ll want to return the order of these back the way they were.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 2]
```diff
@@ -96,21 +110,17 @@ public void onClassificationChange(Location l, NetworkComponent from, NetworkCom
 
         if (component != null) {
             switch (component.getEnergyComponentType()) {
-                case CAPACITOR:
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:99]: It’s probably fine to keep this in, but in the future try to keep the scope of you changes limited to the change itself. Other things like this should be in a separate PR. don’t change this now though unless someone else has an issue with this.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 3]
```diff
@@ -261,10 +271,10 @@ private int tickAllCapacitors() {
     private void updateHologram(@Nonnull Block b, double supply, double demand) {
         if (demand > supply) {
             String netLoss = NumberUtils.getCompactDouble(demand - supply);
-            updateHologram(b, "&4&l- &c" + netLoss + " &7J &e\u26A1");
+            updateHologram(b, "&4&l- &c" + netLoss + " &7J &e⚡");
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:274]: Not really sure I like this change? Can’t say I even have a reason to voice, however.

**Review by Sefiraat — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java (hunk 3 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:274]: I did this for consistency as I saw it like this in some other places, but upon further looking it looks like that was caused by my decompiler just replacing it automatically. Will revert 👍

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java (hunk 2 above)
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java]: No need to remove, it's a good change. I was stating it for the future. Now it's done may as well keep it in :)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 4]
```diff
@@ -59,6 +60,22 @@ public int getRange() {
...
+    }
+    
+    @Nonnull
+    public Map<Location, EnergyNetComponent> getConsumers() {
+        return Collections.unmodifiableMap(consumers);
+    }
+
+    @Nonnull
     @Override
     public String getId() {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:80]: Formatting. Also API methods should always have javadocs.
```suggestion
    public @Nonnull Map<Location, EnergyNetProvider> getGenerators() {
        return Collections.unmodifiableMap(generators);
    }
    
    public @Nonnull Map<Location, EnergyNetComponent> getCapacitors() {
        return Collections.unmodifiableMap(capacitors);
    }
    
    public @Nonnull Map<Location, EnergyNetComponent> getConsumers() {
        return Collections.unmodifiableMap(consumers);
    }

    @Override
    public @Nonnull String getId() {
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by JustAHuman-xD** [comment]
Will do the java docs rn 👍 Wasn't sure if it was necessary or not as it's just a getter 😅

**Comment by JustAHuman-xD** [comment]
Also for that formatting I went off of how it was already formatted. If you look at the bottom of the class the methods have the annotations on separate lines there already. Should I change that to be consistent?

**Comment by JustAHuman-xD** [comment]
I can only request a review from one person at a time, huh, didn't realize it worked that way.

**Comment by Sefiraat** [comment]
> I can only request a review from one person at a time, huh, didn't realize it worked that way.

If a person is assigned (walshy assign me and cookie self-assigned) you can probably leave re-requesting reviews unless some time has passed or you wanna poke us. We get notified without it :)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 5]
```diff
@@ -59,15 +59,32 @@ protected EnergyNet(@Nonnull Location l) {
     public int getRange() {
         return RANGE;
     }
-
+    
+    /**
+     * @return An immutable {@link HashMap}
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:64]: Returns go after the description

**Comment by Sefiraat** [comment]
> Also for that formatting I went off of how it was already formatted. If you look at the bottom of the class the methods have the annotations on separate lines there already. Should I change that to be consistent?

Personally I would leave them for now. A quick search has shown this inconsistency is in many other places in the code base so could be a little project for someone to tidy within another PR

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java (hunk 5 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:64]: Oh I see, my bad uno momento

**Comment by JustAHuman-xD** [comment]
> > I can only request a review from one person at a time, huh, didn't realize it worked that way.
> 
> If a person is assigned (walshy assign me and cookie self-assigned) you can probably leave re-requesting reviews unless some time has passed or you wanna poke us. We get notified without it :)

Ah gotcha, will keep in mind for the future 👍

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java (hunk 5 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java]: Changed it! Also reworded a little bit to hopefully be easier to understand.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 6]
```diff
@@ -58,9 +59,41 @@ protected EnergyNet(@Nonnull Location l) {
...
+    public @Nonnull Map<Location, EnergyNetProvider> getGenerators() {
+        return Collections.unmodifiableMap(generators);
+    }
+    
+    /**
+     * This creates an Immutable Version of {@link #capacitors}.
+     * Using a Key of {@link Location} to a Value of {@link EnergyNetProvider}.
+     * This Map contains all the {@link EnergyNetComponent} with the {@link EnergyNetComponentType#CAPACITOR}
+     * in this Energy Network
+     * @return {@link Map}
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:78]:
```suggestion
     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CAPACITOR} {@link EnergyNetComponent}s.
     *
     * @return An immutable {@link Map} of capacitors
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 7]
```diff
@@ -58,9 +59,41 @@ protected EnergyNet(@Nonnull Location l) {
...
+    public @Nonnull Map<Location, EnergyNetComponent> getCapacitors() {
+        return Collections.unmodifiableMap(capacitors);
+    }
+    
+    /**
+     * This creates an Immutable Version of {@link #consumers}.
+     * Using a Key of {@link Location} to a Value of {@link EnergyNetProvider}.
+     * This Map contains all the {@link EnergyNetComponent} with the {@link EnergyNetComponentType#CONSUMER}
+     * in this Energy .
+     * @return {@link Map}
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:89]:
```suggestion
     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CONSUMER} {@link EnergyNetComponent}s.
     *
     * @return An immutable {@link Map} of consumers
```

**Review by WalshyDev — DISMISSED** [review]
Would be nice to have a unit test but LGTM

**Review by Sefiraat — DISMISSED** [review]
👍

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
You might want to specify "[...] within this EnergyNet instance" on all these methods to avoid people mistaking it for a list of ALL loaded capacitors.

Remember, if this is supposed to be API, we need to be as clear as possible and possibly maintain it for years to come.

**Comment by JustAHuman-xD** [comment]
I had something like that before Walshy committed his changes lol, I'll re add a small thing about it.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 8]
```diff
@@ -58,9 +59,36 @@ protected EnergyNet(@Nonnull Location l) {
     public int getRange() {
         return RANGE;
     }
+    
+    /**
+     * This creates an immutable {@link Map} of {@link EnergyNetProvider}s within this EnergyNet instance.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:64]:
```suggestion
     * This creates an immutable {@link Map} of {@link EnergyNetProvider}s within this {@link EnergyNet} instance.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 9]
```diff
@@ -58,9 +59,36 @@ protected EnergyNet(@Nonnull Location l) {
...
+     *
+     * @return An immutable {@link Map} of generators
+     */
+    public @Nonnull Map<Location, EnergyNetProvider> getGenerators() {
+        return Collections.unmodifiableMap(generators);
+    }
+    
+    /**
+     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CAPACITOR} {@link EnergyNetComponent}s within this EnergyNet instance.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:73]:
```suggestion
     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CAPACITOR} {@link EnergyNetComponent}s within this {@link EnergyNet} instance.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java [hunk 10]
```diff
@@ -58,9 +59,36 @@ protected EnergyNet(@Nonnull Location l) {
...
+     *
+     * @return An immutable {@link Map} of capacitors
+     */
+    public @Nonnull Map<Location, EnergyNetComponent> getCapacitors() {
+        return Collections.unmodifiableMap(capacitors);
+    }
+    
+    /**
+     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CONSUMER} {@link EnergyNetComponent}s within this EnergyNet instance.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/energy/EnergyNet.java:82]:
```suggestion
     * This creates an immutable {@link Map} of {@link EnergyNetComponentType#CONSUMER} {@link EnergyNetComponent}s within this {@link EnergyNet} instance.
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
small nitpick at this point but always use `@link`

**Review by TheBusyBiscuit — APPROVED** [review]
> before Walshy committed his changes lol

*sighs*

> I'll re add a small thing about it.

Thanks, lgtm!

**Review by Sefiraat — APPROVED** [review]
👍 👍

**Pull request merged**